cpp/src/reductions/segmented/valid_counts.hpp

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+namespace cudf {
+class column_device_view;
+
+namespace reduction {
+namespace detail {
+
+/**
+ * @brief Compute the number of valid entries per segment
+ *
+ * If `null_handling == null_policy::EXCLUDE`, the valid count for each
+ * segment omits any null entries. Otherwise, this returns the number
+ * of elements in each segment.
+ *
+ * @param null_mask Null values over which the segment offsets apply
+ * @param has_nulls True if d_col contains any nulls
+ * @param offsets Indices to segment boundaries
+ * @param null_handling How null entries are processed within each segment
+ * @param stream Used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ */
+rmm::device_uvector<size_type> segmented_valid_counts(bitmask_type const* null_mask,
```

Comment:
I'm a bit confused. So when `null_handling != null_policy::EXCLUDE`, this is just a size count, isn't it? If so then the name `_valid_counts` is a bit misleading.

Reply:
Other comment: This can also be adopted in sort-based groupby for `COUNT_VALID` and `COUNT_ALL`.

Reply:
It returns number of valid rows in each segment in both cases.

Reply:
But if `has_null == false || null_handling != null_policy::EXCLUDE` I see it just does this:
```
 rmm::device_uvector<size_type> valid_counts(num_segments, stream, mr);
  thrust::adjacent_difference(
    rmm::exec_policy(stream), offsets.begin() + 1, offsets.end(), valid_counts.begin());
```

Reply:
My confusion comes from when nulls are considered as `INCLUDE`. Even in such cases, nulls are still "invalid elements", right?

Reply:
Ultimately, that is up to the caller how to treat those elements. This is just providing the appropriate count based on the parameters.

Reply:
Valid counts should always count set bits in the null mask, and only use segment lengths via adjacent difference if the column has no nulls / is non-nullable. Counting valid elements does not depend on a “null policy” at all. This parameter should be removed.

Reply:
Is this function’s special behavior with null handling used more than once? (I can’t count the number of invocations right now on mobile.) I would accept “segment counts” but am unsure if factoring out the null handling is highly reusable.

Reply:
Yes, it is used for computing the values (e.g. denominator(s)) for multiple compound aggregations. It is also used for fixed-point product aggregation to help compute the scale size for the output data type. All usages are in this PR. I'm happy with `segmented_counts()` which is more consistent with other utilities here.

Reply:
`segment_counts` sounds good to me.